das/manager.go

```diff
@@ -0,0 +1,190 @@
+package das
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/hashicorp/go-multierror"
+)
+
+// samplingManager coordinates sampling workers and stores current state
+type samplingManager struct {
+	concurrency int
+	fetchFn     fetchFn
+	storeState  storeFn
+
+	state state
+
+	resultCh chan result // fan-in sampling results from worker to coordinator
+	updMaxCh chan uint64 // signals to update max known header height
+	stateCh  chan state  // communicates with backgroundStore routine
+
+	workersWg           sync.WaitGroup
+	workersDone         chan struct{}
+	coordinatorDone     chan struct{}
+	backgroundStoreDone chan struct{}
+}
+
+type fetchFn func(context.Context, uint64) error
+type storeFn func(context.Context, checkpoint)
+
+type result struct {
+	job
+	failed []uint64
+	err    *multierror.Error
+}
+
+func newSamplingManager(
+	concurrency int,
+	fetch fetchFn,
+	storeFn storeFn) *samplingManager {
+	return &samplingManager{
+		concurrency:         concurrency,
+		fetchFn:             fetch,
+		storeState:          storeFn,
+		state:               state{},
+		resultCh:            make(chan result),
+		updMaxCh:            make(chan uint64),
+		stateCh:             make(chan state),
+		workersDone:         make(chan struct{}),
+		coordinatorDone:     make(chan struct{}),
+		backgroundStoreDone: make(chan struct{}),
+	}
+}
+
+func (sm *samplingManager) runCoordinator(ctx context.Context) {
+	var done bool
+
+	for {
+		// launch workers if there are jobs available and concurrency limit is not reached
+		for !done && len(sm.state.inProgress) < sm.concurrency {
+			done = sm.runWorker(ctx)
+		}
+
+		select {
+		case max := <-sm.updMaxCh:
+			if sm.state.updateMaxKnown(max) && done {
+				// unlock workers for newly received headers
+				done = false
+				sm.state.checkDone()
+			}
+		case res := <-sm.resultCh:
+			sm.state.handleResult(res)
+			sm.state.checkDone()
+		case sm.stateCh <- sm.state:
+		case <-ctx.Done():
+			close(sm.coordinatorDone)
+			sm.workersWg.Wait()
+			close(sm.workersDone)
+			return
+		}
+	}
+}
+
+func (sm *samplingManager) runWorker(ctx context.Context) (done bool) {
+	next, done := sm.state.nextJob()
+	if done {
+		// if nothing to sample, no need to start worker
+		return
+	}
+
+	w := newWorker(sm.resultCh, sm.fetchFn)
+	sm.state.putInProgress(next.id, w)
+
+	sm.workersWg.Add(1)
+	go func() {
+		defer sm.workersWg.Done()
+		w.run(ctx, next)
+	}()
+	return false
+}
+
+func (sm *samplingManager) finished(ctx context.Context) error {
+	// wait for coordinator to exit and store state
+	select {
+	case <-sm.coordinatorDone:
+		sm.storeState(ctx, sm.state.toCheckPoint())
```

Comment:
The coordinator itself can do this op. This way we can avoid coordinatorDone and its handling

Reply:
Well, actually, the `storeState` on stop is more correct to be done by the ctx given to `Stop` as it is right now, so no need to change this

Reply:
Still, we should store the checkpoint when workers are done

Reply:
Yep, you got it right. `store` function  is called from `Stop()` to use with `stopCtx` instead of `runCtx` of the coordinator.

Added extra call for `store` after workers exit as you suggested. There was a possibility they progressed 1 more header ahead.